cellcounter/cc_kapi/views.py

```diff
@@ -1,51 +1,243 @@
 from rest_framework import generics
 from rest_framework.response import Response
-from rest_framework.permissions import IsAuthenticated
+from rest_framework.permissions import IsAuthenticatedOrReadOnly, IsAuthenticated
+from rest_framework.exceptions import NotFound, ParseError, NotAuthenticated
+from rest_framework import status
 
-from .models import Keyboard
-from .serializers import KeyboardSerializer
-from .defaults import DEFAULT_KEYBOARD_MAP
+from .models import Keyboard, DefaultKeyboards
+from .serializers import KeyboardSerializer, KeyboardListItemSerializer
 
+from rest_framework import viewsets
 
-class DefaultKeyboardView(generics.RetrieveAPIView):
-    """This returns either the DEFAULT_KEYBOARD_MAP, or the user's
-        primary keyboard"""
-    serializer_class = KeyboardSerializer
+from rest_framework.decorators import detail_route
 
-    def get_object(self):
-        return Keyboard.objects.get(user=self.request.user, is_primary=True)
+from django.core.urlresolvers import reverse
+
+import itertools
+
+class KeyboardViewSet(viewsets.ViewSet):
+    """
+    A ViewSet for manipulating keyboards.
+    """
+
+    permission_classes = (IsAuthenticatedOrReadOnly,)
+
+    _device_type = None
+
+    def device_type_display(self):
+        return Keyboard.DEVICE_TYPES[self._device_type-1][1]
+
+    @property
+    def device_type(self):
+        return self._device_type
+
+    @device_type.setter
+    def device_type(self, val):
+        self._device_type = val
+
+    def _set_user_default_keyboard(self, user, keyboard):
+        """Helper function to set the user's default keyboard.
+        """
+
+        # update or create a default on the user's profile as appropriate
+        if not hasattr(user, 'defaultkeyboards'):
+            defaultkeyboards = DefaultKeyboards.objects.create(user=user)
+        else:
+            defaultkeyboards = user.defaultkeyboards
+
+        if self.device_type == Keyboard.DESKTOP:
+            defaultkeyboards.desktop = keyboard
+        elif self.device_type == Keyboard.MOBILE:
+            defaultkeyboards.mobile = keyboard
+
+        defaultkeyboards.save()
+
+    def _clear_user_default_keyboard(self, user):
+        if not hasattr(user, 'defaultkeyboards'):
+            return
+        defaultkeyboards = user.defaultkeyboards
+
+        if self.device_type == Keyboard.DESKTOP:
+            defaultkeyboards.desktop = None
+        elif self.device_type == Keyboard.MOBILE:
+            defaultkeyboards.mobile = None
+
+        defaultkeyboards.save()
+
+
+    def list(self, request):
```

Comment:
I think we need to rethink how the default flag is handled for keyboards. I think the current solution leads to more work in Python, and more hits to the database than is really necessary. 

My other concern is the idea of built-in objects in the database, how they get there, are updated, and how duplication is prevented (this current code does not robustly handle the potential existence of multiple built-in keyboards for a single device type – see the `retrieve()` method for an example).

I think the default flag should be set in a single place, and shouldn’t be dynamically generated in python. You can access what keyboard the user has set as default via a `KeyboardPreferences` model (a one-to-one with the `User` object), and an API can be exposed to the front end to access this information for logged-in users.

If you’re trying to list all keyboards available to the user, you can then simply do a single query  (using [Q](https://docs.djangoproject.com/en/1.10/topics/db/queries/#complex-lookups-with-q) objects) for each option, rather than post-select processing the queryset in Python:

1. No device type, Authenticated
2. No device type, Unauthenticated
3. Device Type, Authenticated
4. Device Type, Unauthenticated

```
device_type = None

if not device_type:
    if self.request.user.is_authenticated():
        user = self.request.user
        available_keyboards = Keyboard.objects.filter(Q(user=None)|Q(user=user))
    else:
        available_keyboards = Keyboard.objects.filter(user=None)
else:
    if self.request.user.is_authenticated():
        user = self.request.user
        available_keyboards = Keyboard.objects.filter(Q(user=None, device_type=device_type)|Q(user=user, device_type=device_type))
    else:
        available_keyboards = Keyboard.objects.filter(user=None, device_type=device_type)

keyboard_list = KeyboardListItemSerializer(keyboards, many=True)
return Response(keyboard_list.data)
```

So, two issues to consider here:

1. If we store built-in model instances within the database, how they are created and updated, how duplicate instances creation is prevented, and how the code anticipates this risk and handles it if it occurs (read: at least, it needs to log and alert to this happening). What is the best way to approach this thorny issue?
2. Default flag handling. This needs to be thought about a bit more to give a cleaner solution – I feel there is currently too much post-database processing in Python